### PR TITLE
Clean up & fix the alloc_sector count when using a slightly bigger bl…

### PR DESCRIPTION
**Pull request by SeppoTakalo**
…ock.

@AnttiKauppila

### CODE REVIEW

**Comment by AnttiKauppila** [comment]
Re visited some code and refactored couple of things.
Commented out code was left intentionally if there are some findings.

@SeppoTakalo @juhhei01 Please review this.

**Comment by AnttiKauppila** [comment]
smoketests passed

**Comment by juhhei01** [comment]
What is test coverage?

**Comment by AnttiKauppila** [comment]
96.4% I suspect that it is dead code, but we can check that later

File: source/nsdynmemLIB/nsdynmemLIB.c [hunk 1]
```diff
@@ -133,14 +133,15 @@ static int heap_alloc_internal_check(int16_t alloc_size)
 
     if (alloc_size) {
         //
-        temp_sector = alloc_size;
-        temp_int = (temp_sector % temp_size);
-        if (temp_int) {
-            temp_sector += (temp_size - temp_int);
-        }
+        temp_sector = (alloc_size+temp_size-1)/temp_size;
```
Comment (kjbracey-arm) [inline source/nsdynmemLIB/nsdynmemLIB.c]: Maths looks correct.
Lose the "temp_size" - just use "sizeof(int)" - it will be clearer.
Remove the commented code.
Get the formatting right - spaces around operators.

File: source/nsdynmemLIB/nsdynmemLIB.c [hunk 2]
```diff
@@ -133,14 +133,15 @@ static int heap_alloc_internal_check(int16_t alloc_size)
...
+//        temp_int = (temp_sector % temp_size);
+//        if (temp_int) {
+//            temp_sector += (temp_size - temp_int);
+//        }
 
 
-        temp_sector /= temp_size;
+//        temp_sector /= temp_size;
 
         temp_int = (heap_size / temp_size);
```
Comment (kjbracey-arm) [inline source/nsdynmemLIB/nsdynmemLIB.c]: Can you tidy this too? It's basically "if size > (heap_size - 2 words), not valid". Could do that at the top of the function in bytes next to the other size check, rather than in words.

File: source/nsdynmemLIB/nsdynmemLIB.c [hunk 3]
```diff
@@ -155,49 +156,58 @@ static int heap_alloc_internal_check(int16_t alloc_size)
...
-            sector_start -= -(size_end);
-        } else {
-            sector_start -= size_end;
-        }
-        size_start = *sector_start;
-    }
-    if (size_start == size_end) {
+    int size_start;
+//    int size_end;
+    int multiplier = direction?1:-1;
```
Comment (kjbracey-arm) [inline source/nsdynmemLIB/nsdynmemLIB.c]: As it's an internal function, why not make the callers pass -1 or +1, save the extra work.

File: source/nsdynmemLIB/nsdynmemLIB.c [hunk 4]
```diff
@@ -155,49 +156,58 @@ static int heap_alloc_internal_check(int16_t alloc_size)
...
+    size_start = *sector_start;
+    sector_start += multiplier;
+    multiplier = size_start<0?-multiplier:multiplier;
```
Comment (kjbracey-arm) [inline source/nsdynmemLIB/nsdynmemLIB.c]: Overly complicated, and badly formatted. Flipping the "direction" variable based on the free status is weird.

You're just wanting to get the size of the block (independent of free status), which is abs(size_start).
I suggest (assuming direction modified to be +1 or -1):

    size = abs(*sector_start);
    sector_start += (1 + size + 1) * direction;

Which I think is fairly self-documenting. (Could someone maybe add a comment somewhere actually explaining the heap format?)

That does assume multiply-by-variable is fast though. I guess it should be on anything except poxy M0 devices. If we cared about slow-multiply devices, then it could be:

    size = abs(*sector_start);
    if (direction > 0) {
            sector_start += 1 + size + 1;
    } else {
            sector_start -= 1 + size + 1;
    }

File: source/nsdynmemLIB/nsdynmemLIB.c [hunk 5]
```diff
@@ -155,49 +156,58 @@ static int heap_alloc_internal_check(int16_t alloc_size)
...
 }
 #endif
 
-void *ns_dyn_mem_alloc(int16_t alloc_size)
+static void *ns_dyn_mem_internal_alloc(int16_t alloc_size, bool dir_up)
 {
-#ifndef STANDARD_MALLOC
-
-    int *ptr = heap_main_end;
+    int *ptr = dir_up?heap_main:heap_main_end;
```
Comment (kjbracey-arm) [inline source/nsdynmemLIB/nsdynmemLIB.c]: Like the concept, but suggest passing +1/-1 in the first place, if you're going to do the multiplication thing.

File: source/nsdynmemLIB/nsdynmemLIB.c [hunk 6]
```diff
@@ -207,45 +217,60 @@ void *ns_dyn_mem_alloc(int16_t alloc_size)
...
                             //Set sector start len & Move pointer to Allocated data sector
                             *ptr = alloc_sector;
-                            //Move Pointer  to next free place
-                            ptr -= alloc_sector;
-                            //Set return Pointer
-                            retval = (void *) ptr;
-                            ptr--;
-                            //Set End length
-                            *ptr-- = alloc_sector;
+                            if (dir_up) {
```
Comment (kjbracey-arm) [inline source/nsdynmemLIB/nsdynmemLIB.c]: Seems a bit laboured, especially compared to the "hole"-creating bit below. 

Can't you actually unify the two cases anyway - at the minute it's 

    if (making hole) {
          make block
          make hole
    } else {
          make block
    }

Would make sense to avoid repeating the block creation. Can't quite write it out, but something like

   if (space for hole) {
         create hole (now at front, rather than back, because it's easier)
    }
   else {
        increase alloc_sector to take extra space
   }
   create block

File: source/nsdynmemLIB/nsdynmemLIB.c [hunk 7]
```diff
@@ -207,45 +217,60 @@ void *ns_dyn_mem_alloc(int16_t alloc_size)
...
+
                             //Now set new slice start & End len fields
                             //Move pointer to New Free Sector
                             //Calculate new sector size
                             s_size = (s_size - alloc_sector - 2);
 
-                            *ptr-- = -(s_size);
-                            ptr -= s_size;
+                            *ptr = -(s_size);
+                            ptr += multiplier*(s_size+1);
```
Comment (kjbracey-arm) [inline source/nsdynmemLIB/nsdynmemLIB.c]: Where did the +1 come from here?

File: source/nsdynmemLIB/nsdynmemLIB.c [hunk 8]
```diff
@@ -257,7 +282,7 @@ void *ns_dyn_mem_alloc(int16_t alloc_size)
 
 
         if (mem_stat_info_ptr) {
-            alloc_size = (alloc_sector * sizeof(int));
+            alloc_size = ((alloc_sector+2) * sizeof(int));
```
Comment (kjbracey-arm) [inline source/nsdynmemLIB/nsdynmemLIB.c]: Don't think this works - alloc_sector reaches here as 0 if the allocation fails (I think) so this would increase allocation on failure.

File: source/nsdynmemLIB/nsdynmemLIB.c [hunk 9]
```diff
@@ -270,6 +295,97 @@ void *ns_dyn_mem_alloc(int16_t alloc_size)
         platform_exit_critical();
     }
     return retval;
+}
+
+void *ns_dyn_mem_alloc(int16_t alloc_size)
+{
+#ifndef STANDARD_MALLOC
+    return ns_dyn_mem_internal_alloc(alloc_size, false);
+//    int *ptr = heap_main_end;
```
Comment (kjbracey-arm) [inline source/nsdynmemLIB/nsdynmemLIB.c]: Don't leave the old code.

**Comment by AnttiKauppila** [comment]
I have gone through Kevin's comments and also refactored the code even more to be more simple.
@SeppoTakalo @kjbracey-arm @juhhei01 Please re-review this

File: source/nsdynmemLIB/nsdynmemLIB.c [hunk 10]
```diff
@@ -117,77 +115,51 @@ void dev_stat_update(mem_stat_update_t type, int16_t size)
     }
 }
 
-//Fix return value allways good fo aligment
+//Fix return value to be always good for alignment
```
Comment (kjbracey-arm) [inline source/nsdynmemLIB/nsdynmemLIB.c]: Bit misleading - it returns number of sectors, rather than an aligned byte size. Although that's far more obvious from the clearer code anyway.

File: source/nsdynmemLIB/nsdynmemLIB.c [hunk 11]
```diff
@@ -117,77 +115,51 @@ void dev_stat_update(mem_stat_update_t type, int16_t size)
...
 static int heap_alloc_internal_check(int16_t alloc_size)
 {
-    int temp_int, temp_sector = 0, temp_size;
-    temp_size = sizeof(int);
+    int temp_sector = 0;
```
Comment (kjbracey-arm) [inline source/nsdynmemLIB/nsdynmemLIB.c]: Get rid of this silly "name all local variables as temp-something" convention.

File: source/nsdynmemLIB/nsdynmemLIB.c [hunk 12]
```diff
@@ -117,77 +115,51 @@ void dev_stat_update(mem_stat_update_t type, int16_t size)
...
+    } else if (alloc_size > (heap_size - 2 * sizeof(int)) ) {
+        heap_failure(NS_DYN_MEM_ALLOCATE_SIZE_NOT_VALID);
+    } else if (alloc_size) {
+        temp_sector = (alloc_size + sizeof(int) - 1) / sizeof(int);
     }
     return temp_sector;
 }
 
+// Checks that sector length indicators equals
+// Sector has format: Length (word) | data area in words | Length (word)
```
Comment (kjbracey-arm) [inline source/nsdynmemLIB/nsdynmemLIB.c]: What does "sector" actually mean in this code? I was reading it as "word", up to this comment, thinking "alloc_sector" meant "allocation size in words", but I guess it's actually "block"? Can we change it to "block"? A "sector" is usually fixed-size - it's misleading.

Other notes on the format: both lengths are negated if free, and the stored length is the size of the data area in words, not the size of the block.

File: source/nsdynmemLIB/nsdynmemLIB.c [hunk 13]
```diff
@@ -117,77 +115,51 @@ void dev_stat_update(mem_stat_update_t type, int16_t size)
...
-        size_start = *sector_start++;
-        if (size_start < 0) {
-            sector_start += -(size_start);
-        } else {
-            sector_start += size_start;
-        }
-        size_end = *sector_start;
+    int size_start = *sector_start;
+    if (direction > 0) {
+        sector_start += (1 + abs(size_start));
```
Comment (kjbracey-arm) [inline source/nsdynmemLIB/nsdynmemLIB.c]: Incrementing sector_start is weird. Because it won't point to the start any more. Make a new variable.

    sector_end = sector_start + (1 + abs(size_start));

File: source/nsdynmemLIB/nsdynmemLIB.c [hunk 14]
```diff
@@ -117,77 +115,51 @@ void dev_stat_update(mem_stat_update_t type, int16_t size)
...
-            sector_start -= -(size_end);
-        } else {
-            sector_start -= size_end;
-        }
-        size_start = *sector_start;
+        sector_start -= (1 + abs(size_start));
     }
-    if (size_start == size_end) {
+
+    if (size_start && size_start == *sector_start) {
```
Comment (kjbracey-arm) [inline source/nsdynmemLIB/nsdynmemLIB.c]: Why "if size_start"?

File: source/nsdynmemLIB/nsdynmemLIB.c [hunk 15]
```diff
@@ -117,77 +115,51 @@ void dev_stat_update(mem_stat_update_t type, int16_t size)
...
 #endif
 
-void *ns_dyn_mem_alloc(int16_t alloc_size)
+// For direction, use 1 for direction up and -1 for down
+static void *ns_dyn_mem_internal_alloc(int16_t alloc_size, int direction)
 {
-#ifndef STANDARD_MALLOC
-
-    int *ptr = heap_main_end;
+    int *ptr = direction > 0?heap_main:heap_main_end;
```
Comment (kjbracey-arm) [inline source/nsdynmemLIB/nsdynmemLIB.c]: Formatting.

File: source/nsdynmemLIB/nsdynmemLIB.c [hunk 16]
```diff
@@ -117,77 +115,51 @@ void dev_stat_update(mem_stat_update_t type, int16_t size)
...
+    int *end = direction > 0?heap_main_end:heap_main;
+    int *temp_ptr;
```
Comment (kjbracey-arm) [inline source/nsdynmemLIB/nsdynmemLIB.c]: Don't declare this outside the block it's used in.

File: source/nsdynmemLIB/nsdynmemLIB.c [hunk 17]
```diff
@@ -196,67 +168,43 @@ void *ns_dyn_mem_alloc(int16_t alloc_size)
...
+                        /*found block*/
                         if (s_size > (alloc_sector + 4)) {
-                            //debug("Bigger\r\n");
-                            //Set sector start len & Move pointer to Allocated data sector
-                            *ptr = alloc_sector;
-                            //Move Pointer  to next free place
-                            ptr -= alloc_sector;
-                            //Set return Pointer
+                            //There is enough room for new hole so create it first
+                            temp_ptr = (ptr + ((alloc_sector + 2) * direction));
```
Comment (kjbracey-arm) [inline source/nsdynmemLIB/nsdynmemLIB.c]: "hole_ptr" would be a better name. Or maybe "free_ptr". Make them match, anyway.

File: source/nsdynmemLIB/nsdynmemLIB.c [hunk 18]
```diff
@@ -196,67 +168,43 @@ void *ns_dyn_mem_alloc(int16_t alloc_size)
...
+                            free_size = s_size - alloc_sector - 2;
+                            *temp_ptr = -(free_size);
+                            temp_ptr += direction * (free_size + 1);
```
Comment (kjbracey-arm) [inline source/nsdynmemLIB/nsdynmemLIB.c]: Be consistent about multiplying with direction or not. Do it everywhere or nowhere.

File: source/nsdynmemLIB/nsdynmemLIB.c [hunk 19]
```diff
@@ -196,67 +168,43 @@ void *ns_dyn_mem_alloc(int16_t alloc_size)
...
-                            //debug("Bigger\r\n");
-                            //Set sector start len & Move pointer to Allocated data sector
-                            *ptr = alloc_sector;
-                            //Move Pointer  to next free place
-                            ptr -= alloc_sector;
-                            //Set return Pointer
+                            //There is enough room for new hole so create it first
+                            temp_ptr = (ptr + ((alloc_sector + 2) * direction));
+                            free_size = s_size - alloc_sector - 2;
+                            *temp_ptr = -(free_size);
```
Comment (kjbracey-arm) [inline source/nsdynmemLIB/nsdynmemLIB.c]: Pointless brackets

File: source/nsdynmemLIB/nsdynmemLIB.c [hunk 20]
```diff
@@ -196,67 +168,43 @@ void *ns_dyn_mem_alloc(int16_t alloc_size)
...
-
-                            s_size = (s_size - alloc_sector);
-
-                            *ptr-- = -(s_size);
-                            ptr -= s_size;
-                            *ptr = -(s_size);
-                            if (mem_stat_info_ptr) {
-                                alloc_sector -= 2;
-                            }
+                            ptr += alloc_sector;
```
Comment (kjbracey-arm) [inline source/nsdynmemLIB/nsdynmemLIB.c]: Don't need the (void *) cast.

File: source/nsdynmemLIB/nsdynmemLIB.c [hunk 21]
```diff
@@ -196,67 +168,43 @@ void *ns_dyn_mem_alloc(int16_t alloc_size)
...
                 }
                 //Move sector to next
                 //Allocated sector
                 moved += (s_size + 2);
-                ptr -= 2; //Skip Both of length fileds
-                ptr -= s_size;
-                if (heap_main == ptr) {
-                    break;
-                }
+                ptr += direction*(s_size + 2);
```
Comment (kjbracey-arm) [inline source/nsdynmemLIB/nsdynmemLIB.c]: Formatting

File: source/nsdynmemLIB/nsdynmemLIB.c [hunk 22]
```diff
@@ -266,19 +214,28 @@ void *ns_dyn_mem_alloc(int16_t alloc_size)
 
 
         if (mem_stat_info_ptr) {
-            alloc_size = (alloc_sector * sizeof(int));
+            // If alloc fail needs to collect size of failed alloc,
+            // this code needs to be revisited
+            alloc_size = ((alloc_sector + 2) * sizeof(int));
```
Comment (kjbracey-arm) [inline source/nsdynmemLIB/nsdynmemLIB.c]: Interesting. How do we want failures logged? At present each failure adds "rounded up size". Adding 0 doesn't work - it doesn't affect the stat!

You could just update by 1, but I guess I'd try to roughly retain previous behaviour by returning the /original/ alloc request size.

So move this into the if (ret_val), and let the failure return the /original/ alloc_size passed in.
Better still, don't actually modify the original alloc_size (it's confusing) - just put the expression in the OK stat update.

Maybe worth marking the alloc_size parameter "const" to discourage internal tinkering.

File: source/nsdynmemLIB/nsdynmemLIB.c [hunk 23]
```diff
@@ -266,19 +214,28 @@ void *ns_dyn_mem_alloc(int16_t alloc_size)
...
             }
         }
         platform_exit_critical();
     }
     return retval;
+}
+
+void *ns_dyn_mem_alloc(int16_t alloc_size)
+{
+#ifndef STANDARD_MALLOC
```
Comment (kjbracey-arm) [inline source/nsdynmemLIB/nsdynmemLIB.c]: Could you not have a single STANDARD_MALLOC ifdef inside the internal function? Save having two copies of the "call malloc" code?

File: source/nsdynmemLIB/nsdynmemLIB.c [hunk 24]
```diff
@@ -401,67 +270,38 @@ void *ns_dyn_mem_temporary_alloc(int16_t alloc_size)
...
-
-            cur_len = *cur_sector;
-            cur_len = -cur_len; //Invert
-            cur_sector += (cur_len + 1); //New End pointer
-            prev_len += (cur_len + 2); // SET NEW LEN
-        } else {
-            prev_sector = 0;
+        cur_sector--;
+        if (*cur_sector < 0) {
+            total_size += (2 - *cur_sector);
```
Comment (kjbracey-arm) [inline source/nsdynmemLIB/nsdynmemLIB.c]: Half-inclined to write "2 + (-*cur_sector)" to make clearer the negative size thing, but I guess it's fairly obvious.

File: source/nsdynmemLIB/nsdynmemLIB.c [hunk 25]
```diff
@@ -401,67 +270,38 @@ void *ns_dyn_mem_temporary_alloc(int16_t alloc_size)
...
+    if (end != heap_main_end) {
+        end++;
+        if (*end < 0) {
+            total_size += (2 - *end);
+            end += (1 - *end);
+        }else{
+            end--;
+        }
     }
+    *cur_sector = -total_size;
```
Comment (kjbracey-arm) [inline source/nsdynmemLIB/nsdynmemLIB.c]: Why write to *cur_sector? Either it's the same as start, in which case it's pointless, or it's in the middle of a larger reason - still pointless.

File: source/nsdynmemLIB/nsdynmemLIB.c [hunk 26]
```diff
@@ -493,24 +333,17 @@ void ns_dyn_mem_free(void *block)
...
-            //Check  prev sector
-            //Try Link Next between cur
-            ns_free_and_merge_next_sector(ptr, size);
-            //Try Merge Prev with cur
-            ns_merge_prev_sector(ptr);
+            ns_free_and_merge_with_adjacent_sectors(ptr, size);
             if (mem_stat_info_ptr) {
                 //Update Free Counter
-                size = (size * sizeof(int));
+                size = ((size+2) * sizeof(int));
```
Comment (kjbracey-arm) [inline source/nsdynmemLIB/nsdynmemLIB.c]: Formatting. And excess brackets. As previous, I would prefer just using an expression to actually changing the variable.

Very hard to follow if one variable changes units or fundamental meaning throughout a function.

**Comment by kjbracey-arm** [comment]
Looking a lot better now. Nearly there.

File: source/nsdynmemLIB/nsdynmemLIB.c (hunk 14 above)
Comment (AnttiKauppila) [inline source/nsdynmemLIB/nsdynmemLIB.c]: Size must be non-zero

File: source/nsdynmemLIB/nsdynmemLIB.c (hunk 18 above)
Comment (AnttiKauppila) [inline source/nsdynmemLIB/nsdynmemLIB.c]: I rename free_size to hole_size. It is current block size - used size => no multiplication

File: source/nsdynmemLIB/nsdynmemLIB.c (hunk 25 above)
Comment (AnttiKauppila) [inline source/nsdynmemLIB/nsdynmemLIB.c]: This code was there earlier also, it prevents double deletion of the pointer.

File: source/nsdynmemLIB/nsdynmemLIB.c (hunk 14 above)
Comment (kjbracey-arm) [inline source/nsdynmemLIB/nsdynmemLIB.c]: Ah, got it. Was confused because the test on size was done after using the size.

I'd be inclined to move the size == 0 rejection to the top. Particularly as you probably should also be more strict on the test there before trying to follow advance the pointer.

Two other corruption indications are abs(size_start) being > (heap_size / sizeof(int) -  2),  or size being INT_MIN, which would stop abs() working - check that first.

And when doing numeric comparisons, don't use boolean shorthands. It's clearer to write the comparison with zero, rather than taking advantage of zero having the special meaning "false" - most particularly don't mix implicit boolean tests with comparisons, as this "if" does.

File: source/nsdynmemLIB/nsdynmemLIB.c (hunk 25 above)
Comment (kjbracey-arm) [inline source/nsdynmemLIB/nsdynmemLIB.c]: That's a good reason. Add an explanatory comment.

**Comment by AnttiKauppila** [comment]
@kjbracey-arm Please revisit
smoketests are passing locally

File: source/nsdynmemLIB/nsdynmemLIB.c [hunk 27]
```diff
@@ -117,146 +114,96 @@ void dev_stat_update(mem_stat_update_t type, int16_t size)
     }
 }
 
-//Fix return value allways good fo aligment
 static int heap_alloc_internal_check(int16_t alloc_size)
 {
-    int temp_int, temp_sector = 0, temp_size;
-    temp_size = sizeof(int);
+    int block_count = 0;
```
Comment (kjbracey-arm) [inline source/nsdynmemLIB/nsdynmemLIB.c]: Can we rename this function. It isn't really a "check", apart from as a side effect. It mainly converts requested size to size in words.

     static int convert_allocation_size(int16_t requested_bytes)

or something like that?

And make the return variable "words" or something - block count sounds to me like "number of blocks".

File: source/nsdynmemLIB/nsdynmemLIB.c [hunk 28]
```diff
@@ -117,146 +114,96 @@ void dev_stat_update(mem_stat_update_t type, int16_t size)
...
-
-        temp_int = (heap_size / temp_size);
-        temp_int -= 2;
-        if (temp_sector > temp_int) {
-            heap_failure(NS_DYN_MEM_ALLOCATE_SIZE_NOT_VALID);
-            temp_sector = 0;
-        }
+    } else if (alloc_size > (heap_size - 2 * sizeof(int)) ) {
+        heap_failure(NS_DYN_MEM_ALLOCATE_SIZE_NOT_VALID);
+    } else if (alloc_size) {
```
Comment (kjbracey-arm) [inline source/nsdynmemLIB/nsdynmemLIB.c]: Unneeded if.

File: source/nsdynmemLIB/nsdynmemLIB.c [hunk 29]
```diff
@@ -117,146 +114,96 @@ void dev_stat_update(mem_stat_update_t type, int16_t size)
...
+        block_count = (alloc_size + sizeof(int) - 1) / sizeof(int);
     }
-    return temp_sector;
+    return block_count;
 }
 
-static int8_t ns_sector_validate(int *sector_start, int direction)
+// Checks that block length indicators equals
```
Comment (kjbracey-arm) [inline source/nsdynmemLIB/nsdynmemLIB.c]: "make sense" rather than "equals"?

File: source/nsdynmemLIB/nsdynmemLIB.c [hunk 30]
```diff
@@ -117,146 +114,96 @@ void dev_stat_update(mem_stat_update_t type, int16_t size)
...
+// Block has format: Size of data area in words | data area in words | Size of data area in words
```
Comment (kjbracey-arm) [inline source/nsdynmemLIB/nsdynmemLIB.c]: Not all of those "in words" can mean the same thing. How about

  size [1 word] | data area [abs(size) words] | size [1 word]

File: source/nsdynmemLIB/nsdynmemLIB.c [hunk 31]
```diff
@@ -117,146 +114,96 @@ void dev_stat_update(mem_stat_update_t type, int16_t size)
...
-            sector_start -= -(size_end);
-        } else {
-            sector_start -= size_end;
-        }
-        size_start = *sector_start;
+        end -= (1 + abs(size_start));
     }
-    if (size_start == size_end) {
+
+    if (size_start && size_start == *end) {
```
Comment (kjbracey-arm) [inline source/nsdynmemLIB/nsdynmemLIB.c]: Still don't like this. You've got the nasty boolean + comparison mix, and you're not doing other sanity checks. At a minimum make this "size_start != 0", even if you don't want the other checks.

File: source/nsdynmemLIB/nsdynmemLIB.c [hunk 32]
```diff
@@ -117,146 +114,96 @@ void dev_stat_update(mem_stat_update_t type, int16_t size)
...
         ret_val = 0;
     }
     return ret_val;
 }
 #endif
 
-void *ns_dyn_mem_alloc(int16_t alloc_size)
+#ifdef STANDARD_MALLOC
+#ifdef USE_IAR
```
Comment (kjbracey-arm) [inline source/nsdynmemLIB/nsdynmemLIB.c]: Hmm. Take this out? I don't really believe it can still be required.

Note that it's actually "ifdef STANDARD_MALLOC", so it's not even protecting the real allocator code! When it was originally put in in 2013, it was unconditional or ifndef. Probably working around some error in an earlier version of the real allocator.

File: source/nsdynmemLIB/nsdynmemLIB.c [hunk 33]
```diff
@@ -117,146 +114,96 @@ void dev_stat_update(mem_stat_update_t type, int16_t size)
...
-
-    int *ptr = heap_main_end;
     void *retval = 0;
-    int h_size, alloc_sector, s_size;
-    int moved = 0;
-
-    alloc_sector = heap_alloc_internal_check(alloc_size);
-    if (alloc_sector) {
-        h_size = (heap_size / sizeof(int));
+    int alloc_block = heap_alloc_internal_check(alloc_size);
```
Comment (kjbracey-arm) [inline source/nsdynmemLIB/nsdynmemLIB.c]: data_size rather than alloc_block?

File: source/nsdynmemLIB/nsdynmemLIB.c [hunk 34]
```diff
@@ -117,146 +114,96 @@ void dev_stat_update(mem_stat_update_t type, int16_t size)
...
-                s_size = *ptr;
-                if (s_size == 0) {
-                    heap_failure(NS_DYN_MEM_HEAP_SECTOR_CORRUPTED);
-                    retval = 0;
-                    break;
-                } else if (s_size < 0) {
-                    //Free
-                    if (((-s_size) >= alloc_sector)) {
+            if (ns_block_validate(ptr, direction) == 0) {
+                int block_total_size = *ptr;
```
Comment (kjbracey-arm) [inline source/nsdynmemLIB/nsdynmemLIB.c]: block_data_size. It's NOT the total size, which would include that two end markers.

File: source/nsdynmemLIB/nsdynmemLIB.c [hunk 35]
```diff
@@ -117,146 +114,96 @@ void dev_stat_update(mem_stat_update_t type, int16_t size)
...
-                //Allocated sector
-                moved += (s_size + 2);
-                ptr -= 2; //Skip Both of length fileds
-                ptr -= s_size;
-                if (heap_main == ptr) {
-                    break;
-                }
+
+                moved += (block_total_size + 2);
+                ptr += direction * (block_total_size + 2);
```
Comment (kjbracey-arm) [inline source/nsdynmemLIB/nsdynmemLIB.c]: Still being inconsistent about whether multiply by direction is good or bad. I don't have a firm view, but stick to one style (compare ns_block_validate).

File: source/nsdynmemLIB/nsdynmemLIB.c [hunk 36]
```diff
@@ -266,14 +213,15 @@ void *ns_dyn_mem_alloc(int16_t alloc_size)
 
 
         if (mem_stat_info_ptr) {
-            alloc_size = (alloc_sector * sizeof(int));
+            // If alloc fail needs to collect size of failed alloc,
```
Comment (kjbracey-arm) [inline source/nsdynmemLIB/nsdynmemLIB.c]: Comment no longer needed.

File: source/nsdynmemLIB/nsdynmemLIB.c [hunk 37]
```diff
@@ -290,178 +238,49 @@ void *ns_dyn_mem_alloc(int16_t alloc_size)
 #endif
 }
 
+void *ns_dyn_mem_alloc(int16_t alloc_size)
+{
+    return ns_dyn_mem_internal_alloc(alloc_size, -1);
```
Comment (kjbracey-arm) [inline source/nsdynmemLIB/nsdynmemLIB.c]: Maybe worth a comment, as this is the top-level function people will glance at. "Allocates from end of heap".

File: source/nsdynmemLIB/nsdynmemLIB.c [hunk 38]
```diff
@@ -290,178 +238,49 @@ void *ns_dyn_mem_alloc(int16_t alloc_size)
...
-            end_sector += (next_len + 1);
-            size += (next_len + 2);
+    // Theory of operation: Block is always in form | Len | Data | Len |
+    // So we need to check length of previous (if current not heap start)
+    // and next (if current not heap end) blocks. Negative length means
+    // free memory so we can merge freed block with those.
+
+    int *start = cur_block;
+    int *end = cur_block + size + 1;
+    int total_size = size;
```
Comment (kjbracey-arm) [inline source/nsdynmemLIB/nsdynmemLIB.c]: Same argument against "total" as above. Maybe "data_size" and "merged_data_size"?

File: source/nsdynmemLIB/nsdynmemLIB.c [hunk 39]
```diff
@@ -290,178 +238,49 @@ void *ns_dyn_mem_alloc(int16_t alloc_size)
...
+            end--;
         }
-
-    }
-    //Set new sector start & End length
-    if (prev_sector) {
-        *prev_sector = -(prev_len);
-        *cur_sector = -(prev_len);
     }
+    *cur_block = -total_size;
```
Comment (kjbracey-arm) [inline source/nsdynmemLIB/nsdynmemLIB.c]: Actually, shouldn't this be -data_size (the passed-in-size), technically? Doesn't really matter, but might be less confusing when dissecting crash dumps.

And would make sense to do this FIRST, before starting the merge. (That would also avoid simplify the compiler's job if making the above change, as it wouldn't need to remember data_size).

Maybe it shouldn't even be part of this call - it should be done outside this call, and this should just be a merge. (Which in turn would not bother with a separate size parameter, it would just load from the pointer).

File: source/nsdynmemLIB/nsdynmemLIB.c [hunk 40]
```diff
@@ -493,25 +312,17 @@ void ns_dyn_mem_free(void *block)
         heap_failure(NS_DYN_MEM_DOUBLE_FREE);
     } else if (ptr < heap_main || ptr >= heap_main_end) {
         heap_failure(NS_DYN_MEM_POINTER_NOT_VALID);
-    } else if ((size < 0) || ((ptr + size) >= heap_main_end)) {
+    } else if ((ptr + size) >= heap_main_end) {
         heap_failure(NS_DYN_MEM_POINTER_NOT_VALID);
     } else {
-        // Validate Sector
-        if (ns_sector_validate(ptr, 1) != 0) {
+        // Validate block
```
Comment (kjbracey-arm) [inline source/nsdynmemLIB/nsdynmemLIB.c]: Silly comment. Repeats the function name.

File: source/nsdynmemLIB/nsdynmemLIB.c (hunk 37 above)
Comment (AnttiKauppila) [inline source/nsdynmemLIB/nsdynmemLIB.c]: Already commented in header
